gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeTableOperations.java

```diff
@@ -0,0 +1,227 @@
+      if (currentMetadataLocation() != null) {
+        // Re-throws the exception because the table must exist in this case.
+        throw e;
+      }
+    }
+    refreshFromMetadataLocation(metadataLocation);
+  }
+
+  // The doCommit method should provide implementation on how to update with metadata location
+  // atomically
+  @Override
+  public void doCommit(TableMetadata base, TableMetadata metadata) {
+    boolean isNewTable = base == null;
+    String newMetadataLocation = writeNewMetadataIfRequired(isNewTable, metadata);
+
+    CommitStatus commitStatus = CommitStatus.FAILURE;
+    try {
+      if (isNewTable) {
+        createTable(newMetadataLocation, metadata);
+      } else {
+        updateTable(base.metadataFileLocation(), newMetadataLocation, metadata);
+      }
+      commitStatus = CommitStatus.SUCCESS;
+    } catch (CommitFailedException | CommitStateUnknownException e) {
+      throw e;
+    } catch (Throwable e) {
+      commitStatus = checkCommitStatus(newMetadataLocation, metadata);
+      if (commitStatus == CommitStatus.FAILURE) {
+        throw new CommitFailedException(e, "Failed to commit");
+      }
+      if (commitStatus == CommitStatus.UNKNOWN) {
+        throw new CommitStateUnknownException(e);
+      }
+    } finally {
+      try {
+        if (commitStatus == CommitStatus.FAILURE) {
+          LOG.warn("Failed to commit updates to table {}", tableName());
+          io().deleteFile(newMetadataLocation);
+        }
+      } catch (RuntimeException e) {
+        LOG.error(
+            "Failed to cleanup metadata file at {} for table {}",
+            newMetadataLocation,
+            tableName(),
+            e);
+      }
+    }
+  }
+
+  @Override
+  public String tableName() {
+    return String.format(
+        "%s.%s.%s", tableName.getCatalog(), tableName.getDatabase(), tableName.getTable());
+  }
+
+  @Override
+  public FileIO io() {
+    return fileIO;
+  }
+
+  private void createTable(String newMetadataLocation, TableMetadata metadata) {
+    LOG.debug("Creating a new Iceberg table: {}", tableName());
+    client.createTable(tableName, makeNewTable(metadata, newMetadataLocation));
+  }
+
+  /** Update table properties with concurrent update detection using etag. */
+  private void updateTable(
+      String oldMetadataLocation, String newMetadataLocation, TableMetadata metadata) {
+    Table table = client.getTable(tableName);
+    String etag = table.getEtag();
+    Preconditions.checkArgument(
+        !etag.isEmpty(),
+        "Etag of legacy table %s is empty, manually update the table by BigLake API or recreate and retry",
+        tableName());
+    HiveTableOptions options = table.getHiveOptions();
+
+    // If `metadataLocationFromMetastore` is different from metadata location of base, it means
+    // someone has updated metadata location in metastore, which is a conflict update.
+    String metadataLocationFromMetastore =
+        options.getParametersOrDefault(METADATA_LOCATION_PROP, "");
```

Comment:
This allows the metadata location to be empty in the table. I think it should check for null like we do in the rest of the library. We don't generally fill in an empty string and check for empty when we can check for `null`.

Reply:
Protobuf Java only have getFooOrDefault and getFooOrThrow. I kept getParametersOrDefault and do an empty string check immediately. It is the same as checking nullness. Catching an error or returning a map from proto + get value are more verbose. Please let me know whether it works.

Reply:
I think this should use the `getParamtersOrThrow` option.

Reply:
Done.